addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go

```diff
@@ -0,0 +1,238 @@
+// Copyright YEAR VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/utils/pointer"
+
+	clusterapipatchutil "sigs.k8s.io/cluster-api/util/patch"
+
+	clusterapiutil "sigs.k8s.io/cluster-api/util"
+
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/constants"
+
+	"gopkg.in/yaml.v3"
+
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	clusterapiv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	addonsv1alpha1 "github.com/vmware-tanzu/tanzu-framework/apis/addons/v1alpha1"
+)
+
+// SecretGenControllerConfigReconciler reconciles a SecretGenControllerConfig object
+type SecretGenControllerConfigReconciler struct {
+	Client client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+//SecretGenControllerConfigSpecYaml is used for yaml marshalling for secretGen controller config
+type SecretGenControllerConfigSpecYaml struct {
+	Namespace       string `yaml:"namespace,omitempty"`
+	CreateNamespace bool   `yaml:"createNamespace,omitempty"`
+}
+
+//+kubebuilder:rbac:groups=addons.tanzu.vmware.com,resources=secretgencontrollerconfigs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=addons.tanzu.vmware.com,resources=secretgencontrollerconfigs/status,verbs=get;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *SecretGenControllerConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("secretgencontrollerconfig", req.NamespacedName)
+
+	// get secret-gen config object
+	secretGenConfig := &addonsv1alpha1.SecretGenControllerConfig{}
+	if err := r.Client.Get(ctx, req.NamespacedName, secretGenConfig); err != nil {
+		log.Error(err, "unable to fetch SecretGenControllerConfig")
+		return ctrl.Result{}, err
+	}
+
+	// verify that the cluster related to config is present
+	cluster := &clusterapiv1beta1.Cluster{}
+	if err := r.Client.Get(ctx, req.NamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("Cluster not found")
+			return ctrl.Result{}, nil
+		}
+
+		log.Error(err, "unable to fetch cluster")
+		return ctrl.Result{}, err
+	}
+
+	// Get client for the cluster that we deploy data value secret
+	remoteClient, err := util.GetClusterClient(ctx, r.Client, r.Scheme, clusterapiutil.ObjectKey(cluster))
```

Comment:
Wh do we need to come up with ```remoteClient```? Is this same controller going to eventually workload clusters config too? I mean why not using ```r.client``` itself?

Reply:
Question: `xxxConfig` API exists on mgmt cluster only, using `r.Client` should be sufficient.

Reply:
We need to deploy the data value secret into the workload clusters (remote cluster)

Reply:
AFAIK the secret has to be created in the management/supervisor cluster and xxConfig.status.secretRef set to the secret name. 

TanzuBootStrapCluster controller will copy the secret to workload cluster.

@vijaykatam  @danniel1205  to confirm

Reply:
That is correct, the controller for the individual API types is responsible for creating a secret and adding it to status.